### PR TITLE
feat(rome_js_ananlyze): noImportAssign

### LINKED ISSUE #2923: 📎 noImportAssign
### Description

https://github.com/rome/tools/blob/archived-js/website/src/docs/lint/rules/js/noImportAssign.md

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
## Summary
1. Resolved #2923 

## Test Plan
1. cargo test

### CODE REVIEW

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+use rome_rowan::{declare_node_union, AstNode};
+
+declare_rule! {
+    ///  Disallow assigning to imported bindings
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs:18]: Could you add a case with `import { a } from "b"`?

Comment (ematipico) [reply crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs:18]: Could you add a case with `import a, * as b from "d"` and we re-assign `b`?

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        name: "noImportAssign"
+    }
+}
+
+impl Rule for NoImportAssign {
+    const CATEGORY: RuleCategory = RuleCategory::Lint;
+
+    type Query = Semantic<JsAnyImportLike>;
+    /// The first element of the tuple is the invalid assignment SyntaxNode, the second element of the tuple is the import binding SyntaxNode
+    type State = (JsSyntaxNode, JsSyntaxNode);
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs:62]: Wouldn't it be better to use `JsAnyBinding` in the first element of the tuple? Or even using an union to identify the local names. Using `JsSyntaxNode` is very loose and error prone.

Same thing for the second element of the tuple, we should use an existing AST node or a new union node.

**Review by ematipico — CHANGES_REQUESTED** [review]

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs (hunk 2 above)
Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs]: Done

**Comment by IWANABETHATGUY** [comment]
!bench_analyzer

**Comment by IWANABETHATGUY** [comment]
@xunilrj , would you mind helping with reviewing?

**Comment by IWANABETHATGUY** [comment]
!bench_analyzer

**Comment by IWANABETHATGUY** [comment]
The performance looks good now.

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            JsAnyImportLike::JsNamespaceImportSpecifier(specifier) => specifier.local_name().ok(),
+            // `import a, * as b from 'y'`
+            //         ^
+            JsAnyImportLike::JsDefaultImportSpecifier(specifier) => specifier.local_name().ok(),
+        };
+        local_name_binding
+            .and_then(|binding| {
+                let ident_binding = binding.as_js_identifier_binding()?;
+                let model = ctx.model();
+                for reference in model.all_references(ident_binding) {
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs:94]: You can do ```all_writes``` here. They should be rare.

**Review by xunilrj — APPROVED** [review]

**Review by ematipico — APPROVED** [review]

**Comment by ematipico** [comment]
Let's hold the merging of this PR until this PR https://github.com/rome/tools/pull/2980 lands

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    /// ```js,expect_diagnostic
+    /// import d from "y";
+    /// d += 1;
+    /// ```
+    /// ```js,expect_diagnostic
+    /// import * as e from "y";
+    /// e = 1;
+    /// ```
+    pub(crate) NoImportAssign {
+        version: "0.8.0",
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs:53]: This should be `0.9.0` now

Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/src/semantic_analyzers/js/no_import_assign.rs]: Done

**Comment by IWANABETHATGUY** [comment]
Since https://github.com/rome/tools/pull/2980 has landed, I thinks this pull request is good to merge now.

**Pull request merged**